src/main/java/com/minecolonies/coremod/entity/ai/mobs/AbstractArcherBarbarian.java

```diff
@@ -0,0 +1,302 @@
+            }
+        }
+
+        return livingdata;
+    }
+
+    /**
+     * sets this entity's combat AI.
+     */
+    public void setCombatTask()
+    {
+        if (this.world != null && !this.world.isRemote)
+        {
+            this.tasks.removeTask(this.aiAttackOnCollide);
+            this.tasks.removeTask(this.aiArrowAttack);
+            final ItemStack itemstack = this.getHeldItemMainhand();
+
+            if (itemstack.getItem() == Items.BOW)
+            {
+                int i = BOW_DAMAGE;
+
+                if (this.world.getDifficulty() != EnumDifficulty.HARD)
+                {
+                    i = BOW_DAMAGE_HARD;
+                }
+
+                this.aiArrowAttack.setAttackCooldown(i);
+                this.tasks.addTask(4, this.aiArrowAttack);
+            }
+            else
+            {
+                this.tasks.addTask(4, this.aiAttackOnCollide);
+            }
+        }
+    }
+
+    /**
+     * Attack the specified entity using a ranged attack.
+     */
+    public void attackEntityWithRangedAttack(final EntityLivingBase target, final float distanceFactor)
+    {
+        final EntityArrow entityarrow = this.getArrow(distanceFactor);
+        final double d0 = target.posX - this.posX;
+        final double d1 = target.getEntityBoundingBox().minY + (double) (target.height / 3.0F) - entityarrow.posY;
+        final double d2 = target.posZ - this.posZ;
+        final double d3 = (double) MathHelper.sqrt(d0 * d0 + d2 * d2);
+        entityarrow.setThrowableHeading(d0,
+          d1 + d3 * GRAVITY_MULTIPLIER,
+          d2,
+          ARROW_VELOCITY,
+          (float) (DIFFICULTY_SUBTRACTER - this.world.getDifficulty().getDifficultyId() * DIFFICULTY_MULTIPLIER));
+        this.playSound(SoundEvents.ENTITY_SKELETON_SHOOT, 1.0F, 1.0F / (this.getRNG().nextFloat() * SOUND_MULTIPLIER + SOUND_ADDER));
+        this.world.spawnEntity(entityarrow);
+    }
+
+    protected EntityArrow getArrow(final float p_190726_1_)
+    {
+        final EntityTippedArrow entitytippedarrow = new EntityTippedArrow(this.world, this);
+        entitytippedarrow.setEnchantmentEffectsFromEntity(this, p_190726_1_);
+        return entitytippedarrow;
+    }
+
+    /**
+     * (abstract) Protected helper method to read subclass entity data from NBT.
+     */
+    public void readEntityFromNBT(final NBTTagCompound compound)
+    {
+        super.readEntityFromNBT(compound);
+        this.setCombatTask();
+    }
+
+    public void setItemStackToSlot(final EntityEquipmentSlot slotIn, final ItemStack stack)
+    {
+        super.setItemStackToSlot(slotIn, stack);
+
+        if (!this.world.isRemote && slotIn == EntityEquipmentSlot.MAINHAND)
+        {
+            this.setCombatTask();
+        }
+    }
+
+    public float getEyeHeight()
+    {
+        return ENTITY_EYE_HEIGHT;
+    }
+
+    /**
+     * Returns the Y Offset of this entity.
+     */
+    public double getYOffset()
+    {
+        return -0.6D;
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Assign this magic number 0.6D to a well-named constant, and use the constant instead. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS109)